src/inferencer/__tests__/inferencer.ts

```diff
@@ -0,0 +1,378 @@
+import { mockContext } from '../../mocks/context'
+import { listPrelude } from '../../stdlib/list.prelude'
+import { parse } from '../../parser/parser'
+import { analyse } from '../inferencer'
+import { TypeAnnotatedNode } from '../../types'
+import * as es from 'estree'
+import { typeToString } from '../../utils/stringify'
+import { parseError } from '../../index'
+import { validateAndAnnotate } from '../../validator/validator'
+
+function topLevelTypesToString(program: TypeAnnotatedNode<es.Program>) {
+  return program.body
+    .filter(node => ['VariableDeclaration', 'FunctionDeclaration'].includes(node.type))
+    .map(
+      (
+        node: TypeAnnotatedNode<es.VariableDeclaration> | TypeAnnotatedNode<es.FunctionDeclaration>
+      ) => {
+        const id =
+          node.type === 'VariableDeclaration'
+            ? (node.declarations[0].id as es.Identifier).name
+            : node.id?.name!
+        const type =
+          node.typability === 'Untypable' ? "Couldn't infer type" : typeToString(node.inferredType!)
+        return `${id}: ${type}`
+      }
+    )
+    .join('\n')
+}
+
+test('Type inference of list prelude', async () => {
+  const code = listPrelude
+  const context = mockContext(2)
+  const program = parse(code, context)!
+  expect(program).not.toBeUndefined()
+  validateAndAnnotate(program, context)
+  analyse(program, context)
+  expect(program).toMatchSnapshot()
+  expect(parseError(context.errors)).toMatchInlineSnapshot(`
+"Line 17: A type mismatch was detected in the binary expression:
+  x === y
+The binary operator (===) expected two operands with types:
+  Addable1 === Addable1
+but instead it received two operands of types:
+  [T1, T1] === [T1, T1]
+Line 72: A type mismatch was detected in the function call:
+  list_to_string(tail(xs))
+The function expected an argument of type:
+  T1
+but instead received an argument of type:
+  List<T1>"
+`)
+  expect(topLevelTypesToString(program!)).toMatchInlineSnapshot(`
+"is_list: List<T1> -> T2
+equal: Couldn't infer type
+length: List<T1> -> number
+map: (T1 -> T2, List<T1>) -> List<T2>
+build_list: (number, number -> T1) -> List<T1>
+for_each: (T1 -> T2, List<T1>) -> boolean
+list_to_string: Couldn't infer type
+reverse: List<T1> -> List<T1>
+append: (List<T1>, List<T1>) -> List<T1>
+member: (Addable1, List<Addable1>) -> List<Addable1>
+remove: (Addable1, List<Addable1>) -> List<Addable1>
+remove_all: (Addable1, List<Addable1>) -> List<Addable1>
+filter: (T1 -> boolean, List<T1>) -> List<T1>
+enum_list: (number, number) -> List<number>
+list_ref: (List<T1>, number) -> T1
+accumulate: ((T1, T2) -> T2, T2, List<T1>) -> T2"
+`)
+})
+
+test('Type inference of permutation', async () => {
+  const code = `function accumulate(op, init, xs) {
+    return is_null(xs) ? init : op(head(xs), accumulate(op, init, xs));
```

Comment:
Should be this right?
`accumulate(op, init, tail(xs))`

Reply:
Yeah, thanks for noticing!